File: apps/android/app/src/main/java/pro/wewed/app/services/SessionAuthority.kt
package pro.wewed.app.services

import pro.wewed.app.models.*

/**
 * Accounts that can sign in to a Shadow build. Only COUPLE is backed by a source relationship;
 * every other account is either the guest's own invitation or an explicit UAT overlay.
 */
enum class ShadowAccount(val key: String) {
    COUPLE("couple"),
    PLANNER("planner"),
    COUPLE_AND_PLANNER("couple_planner"),
    GUEST("guest"),
    VENDOR("vendor"),
    USHER("usher"),
    COORDINATOR("coordinator"),
    ADMIN("admin");

    companion object {
        fun fromKey(raw: String?): ShadowAccount? {
            val key = raw?.trim()?.lowercase()?.replace('-', '_') ?: return null
            return entries.firstOrNull { it.key == key }
        }
    }
}

sealed class SessionAuthorityError(message: String) : IllegalStateException(message) {
    data object ProductionAuthenticationUnavailable :
        SessionAuthorityError("Production sign-in is not available in this native build.")
    data object InvitationTokenRequired :
        SessionAuthorityError("A guest can only sign in with their own invitation link.")
    data object InvitationNotRecognised :
        SessionAuthorityError("This invitation link was not recognised.")
    data object NoVendorEngagement :
        SessionAuthorityError("There is no vendor engagement to scope a vendor session to.")
}

/**
 * Resolves who is signing in and exactly which roles they hold for which wedding.
 * UI never picks a role on its own; it can only choose among the grants returned here.
 */
object SessionAuthority {
    const val DEFAULT_GUEST_INVITATION_TOKEN = "shadow-pending-guest"

    private const val PLANNER_OVERLAY_NOTE =
        "Test access only. This planner's interest was accepted, but no planner engagement is recorded for this wedding."
    private const val VENDOR_OVERLAY_NOTE =
        "Test access only. No vendor sign-in is recorded for this wedding; this session is limited to one vendor's own engagement."
    private const val USHER_OVERLAY_NOTE =
        "Test access only. Gate team sign-ins are not recorded for this wedding."
    private const val COORDINATOR_OVERLAY_NOTE =
        "Test access only. No coordinator membership is recorded for this wedding."
    private const val ADMIN_OVERLAY_NOTE =
        "Test access only. Support access is read-only and every view is recorded in the audit log."

    suspend fun signIn(
        account: ShadowAccount,
        environment: NativeDataEnvironment,
        wedding: WeddingRepository,
        planner: PlannerDashboardRepository,
        invitationToken: String? = null
    ): AuthorizedSession {
        if (!environment.allowsMutableNativeDevelopment) {
            throw SessionAuthorityError.ProductionAuthenticationUnavailable
        }
        if (account == ShadowAccount.GUEST) {
            return signInWithInvitation(invitationToken ?: DEFAULT_GUEST_INVITATION_TOKEN, environment, wedding)
        }

        val record = wedding.getWedding()
        fun overlay(role: AppRole, note: String, vendorId: String? = null) = RoleGrant(
            role = role,
            weddingId = record.id,
            weddingTitle = record.coupleNames,
            provenance = GrantProvenance.SHADOW_TEST_OVERLAY,
            provenanceNote = note,
            vendorId = vendorId
        )
        val coupleGrant = RoleGrant(
            role = AppRole.COUPLE,
            weddingId = record.id,
            weddingTitle = record.coupleNames,
            provenance = if (environment == NativeDataEnvironment.PRIVATE_REAL_SHADOW) {
                GrantProvenance.SOURCE_RECORD
            } else {
                GrantProvenance.SANITIZED_FIXTURE
            },
            provenanceNote = "This account owns the wedding record."
        )

        return when (account) {
            ShadowAccount.COUPLE ->
                AuthorizedSession("couple:${record.id}", record.coupleNames, listOf(coupleGrant))
            ShadowAccount.PLANNER ->
                AuthorizedSession("planner:${record.id}", plannerDisplayName(planner), listOf(overlay(AppRole.PLANNER, PLANNER_OVERLAY_NOTE)))
            ShadowAccount.COUPLE_AND_PLANNER ->
                AuthorizedSession(
                    "couple-planner:${record.id}",
                    record.coupleNames,
                    listOf(coupleGrant, overlay(AppRole.PLANNER, PLANNER_OVERLAY_NOTE))
                )
            ShadowAccount.VENDOR -> {
                val engagement = planner.getVendorEngagements().firstOrNull { !it.vendorId.isNullOrBlank() }
                    ?: throw SessionAuthorityError.NoVendorEngagement
                AuthorizedSession(
                    "vendor:${engagement.vendorId}",
                    engagement.vendorName,
                    listOf(overlay(AppRole.VENDOR, VENDOR_OVERLAY_NOTE, vendorId = engagement.vendorId))
                )
            }
            ShadowAccount.USHER ->
                AuthorizedSession("usher:${record.id}", "Gate team", listOf(overlay(AppRole.USHER, USHER_OVERLAY_NOTE)))
            ShadowAccount.COORDINATOR ->
                AuthorizedSession("coordinator:${record.id}", "Wedding-day coordinator", listOf(overlay(AppRole.COORDINATOR, COORDINATOR_OVERLAY_NOTE)))
            ShadowAccount.ADMIN ->
                AuthorizedSession("admin:support", "Wewed support", listOf(overlay(AppRole.ADMIN, ADMIN_OVERLAY_NOTE)))
            ShadowAccount.GUEST -> error("handled above")
        }
    }

    /** A guest proves who they are with their own invitation; the grant is pinned to that guest record. */
    suspend fun signInWithInvitation(
        token: String,
        environment: NativeDataEnvironment,
        wedding: WeddingRepository
    ): AuthorizedSession {
        if (!environment.allowsMutableNativeDevelopment) {
            throw SessionAuthorityError.ProductionAuthenticationUnavailable
        }
        val trimmed = token.trim()
        if (trimmed.isEmpty()) throw SessionAuthorityError.InvitationTokenRequired
        val record = wedding.getWedding()
        val invitation = runCatching { wedding.resolveInvitation(record.id, trimmed) }
            .getOrElse { throw SessionAuthorityError.InvitationNotRecognised }
        val guestId = invitation.guestId ?: throw SessionAuthorityError.InvitationNotRecognised
        return AuthorizedSession(
            accountId = "guest:$guestId",
            displayName = invitation.guestName,
            grants = listOf(
                RoleGrant(
                    role = AppRole.GUEST,
                    weddingId = record.id,
                    weddingTitle = record.coupleNames,
                    provenance = GrantProvenance.INVITATION_TOKEN,
                    provenanceNote = "Signed in with this guest's own invitation.",
                    guestId = guestId
                )
            )
        )
    }

    private suspend fun plannerDisplayName(planner: PlannerDashboardRepository): String =
        planner.getDashboard().plannerContext.substringBefore(" •").trim().ifEmpty { "Planner" }
}

File: apps/ios/Wewed/Services/SessionAuthority.swift
import Foundation

/// Accounts that can sign in to a Shadow build. Only couple is backed by a source relationship;
/// every other account is either the guest's own invitation or an explicit UAT overlay.
public enum ShadowAccount: String, CaseIterable, Sendable {
    case couple
    case planner
    case coupleAndPlanner = "couple_planner"
    case guest
    case vendor
    case usher
    case coordinator
    case admin

    public static func from(key raw: String?) -> ShadowAccount? {
        guard let raw else { return nil }
        let key = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased().replacingOccurrences(of: "-", with: "_")
        return ShadowAccount(rawValue: key)
    }
}

public enum SessionAuthorityError: Error, Equatable, LocalizedError {
    case productionAuthenticationUnavailable
    case invitationTokenRequired
    case invitationNotRecognised
    case noVendorEngagement

    public var errorDescription: String? {
        switch self {
        case .productionAuthenticationUnavailable: return "Production sign-in is not available in this native build."
        case .invitationTokenRequired: return "A guest can only sign in with their own invitation link."
        case .invitationNotRecognised: return "This invitation link was not recognised."
        case .noVendorEngagement: return "There is no vendor engagement to scope a vendor session to."
        }
    }
}

/// Resolves who is signing in and exactly which roles they hold for which wedding.
/// UI never picks a role on its own; it can only choose among the grants returned here.
/// Mirrors Android `SessionAuthority`.
public enum SessionAuthority {
    public static let defaultGuestInvitationToken = "shadow-pending-guest"

    static let plannerOverlayNote =
        "Test access only. This planner's interest was accepted, but no planner engagement is recorded for this wedding."
    static let vendorOverlayNote =
        "Test access only. No vendor sign-in is recorded for this wedding; this session is limited to one vendor's own engagement."
    static let usherOverlayNote = "Test access only. Gate team sign-ins are not recorded for this wedding."
    static let coordinatorOverlayNote = "Test access only. No coordinator membership is recorded for this wedding."
    static let adminOverlayNote = "Test access only. Support access is read-only and every view is recorded in the audit log."

    public static func signIn(
        account: ShadowAccount,
        environment: NativeDataEnvironment,
        wedding: WeddingRepositoryProtocol,
        planner: PlannerDashboardRepositoryProtocol,
        invitationToken: String? = nil
    ) async throws -> AuthorizedSession {
        guard environment.allowsMutableNativeDevelopment else {
            throw SessionAuthorityError.productionAuthenticationUnavailable
        }
        if account == .guest {
            return try await signInWithInvitation(
                token: invitationToken ?? defaultGuestInvitationToken,
                environment: environment,
                wedding: wedding
            )
        }

        let record = try await wedding.getWedding()
        func overlay(_ role: AppRole, _ note: String, vendorId: String? = nil) -> RoleGrant {
            RoleGrant(
                role: role,
                weddingId: record.id,
                weddingTitle: record.coupleNames,
                provenance: .shadowTestOverlay,
                provenanceNote: note,
                vendorId: vendorId
            )
        }
        let coupleGrant = RoleGrant(
            role: .couple,
            weddingId: record.id,
            weddingTitle: record.coupleNames,
            provenance: environment == .privateRealShadow ? .sourceRecord : .sanitizedFixture,
            provenanceNote: "This account owns the wedding record."
        )

        switch account {
        case .couple:
            return AuthorizedSession(accountId: "couple:\(record.id)", displayName: record.coupleNames, grants: [coupleGrant])
        case .planner:
            return AuthorizedSession(
                accountId: "planner:\(record.id)",
                displayName: try await plannerDisplayName(planner),
                grants: [overlay(.planner, plannerOverlayNote)]
            )
        case .coupleAndPlanner:
            return AuthorizedSession(
                accountId: "couple-planner:\(record.id)",
                displayName: record.coupleNames,
                grants: [coupleGrant, overlay(.planner, plannerOverlayNote)]
            )
        case .vendor:
            guard let engagement = try await planner.getVendorEngagements().first(where: { !($0.vendorId ?? "").isEmpty }),
                  let vendorId = engagement.vendorId else {
                throw SessionAuthorityError.noVendorEngagement
            }
            return AuthorizedSession(
                accountId: "vendor:\(vendorId)",
                displayName: engagement.vendorName,
                grants: [overlay(.vendor, vendorOverlayNote, vendorId: vendorId)]
            )
        case .usher:
            return AuthorizedSession(accountId: "usher:\(record.id)", displayName: "Gate team", grants: [overlay(.usher, usherOverlayNote)])
        case .coordinator:
            return AuthorizedSession(accountId: "coordinator:\(record.id)", displayName: "Wedding-day coordinator", grants: [overlay(.coordinator, coordinatorOverlayNote)])
        case .admin:
            return AuthorizedSession(accountId: "admin:support", displayName: "Wewed support", grants: [overlay(.admin, adminOverlayNote)])
        case .guest:
            preconditionFailure("handled above")
        }
    }

    /// A guest proves who they are with their own invitation; the grant is pinned to that guest record.
    public static func signInWithInvitation(
        token: String,
        environment: NativeDataEnvironment,
        wedding: WeddingRepositoryProtocol
    ) async throws -> AuthorizedSession {
        guard environment.allowsMutableNativeDevelopment else {
            throw SessionAuthorityError.productionAuthenticationUnavailable
        }
        let trimmed = token.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { throw SessionAuthorityError.invitationTokenRequired }
        let record = try await wedding.getWedding()
        let invitation: InvitationContext
        do {
            invitation = try await wedding.resolveInvitation(weddingSlug: record.id, token: trimmed)
        } catch {
            throw SessionAuthorityError.invitationNotRecognised
        }
        guard let guestId = invitation.guestId else { throw SessionAuthorityError.invitationNotRecognised }
        return AuthorizedSession(
            accountId: "guest:\(guestId)",
            displayName: invitation.guestName,
            grants: [
                RoleGrant(
                    role: .guest,
                    weddingId: record.id,
                    weddingTitle: record.coupleNames,
                    provenance: .invitationToken,
                    provenanceNote: "Signed in with this guest's own invitation.",
                    guestId: guestId
                )
            ]
        )
    }

    private static func plannerDisplayName(_ planner: PlannerDashboardRepositoryProtocol) async throws -> String {
        let context = try await planner.getDashboard().plannerContext
        let name = context.components(separatedBy: " •").first?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return name.isEmpty ? "Planner" : name
    }
}
